### PR TITLE
Switch to a TypeScript version that is compatible with the used ESLint version.

### PR DESCRIPTION
**Pull request by AugustusKling**
I could not run `yarn run test:lint` as the TypeScript version 5.6.2 seem to be incompatible with the ESLint version 8.57 or some of the used ESLint plugins.

It says complains about not finding some dependencies (log shortended here)
```
src/plugin.ts:21:29 - error TS2307: Cannot find module '@yarnpkg/core' or its corresponding type declarations.

21 import type { Plugin } from '@yarnpkg/core'
                               ~~~~~~~~~~~~~~~

src/plugin.ts:27:28 - error TS4113: This member cannot have an 'override' modifier because it is not declared in the base class 'MakeSbomCommand'.

27   static override readonly paths = [
                              ~~~~~

src/plugin.ts:35:28 - error TS4112: This member cannot have an 'override' modifier because its containing class 'CyclonedxVersionCommand' does not extend another class.

35   static override readonly paths = CyclonedxCommand.paths.map(p => [...p, '--version'])
                              ~~~~~

src/plugin.ts:39:10 - error TS2339: Property 'context' does not exist on type 'CyclonedxVersionCommand'.

39     this.context.stdout.write(`${name} v${version}\n`)
            ~~~~~~~


Found 38 errors in 6 files.

Errors  Files
     3  src/_helpers.ts:20
     1  src/_yarnCompat.ts:20
    10  src/builders.ts:21
    17  src/commands.ts:21
     2  src/logger.ts:20
     5  src/plugin.ts:20
```

Once the TypeScript version is switched, ESLint runs without errors.

### CODE REVIEW

**Comment by jkowalleck** [comment]
for me, locally, it works when run as described  in https://github.com/CycloneDX/cyclonedx-node-yarn/blob/main/CONTRIBUTING.md

> Apply the coding style via:
> `yarn run cs-fix`

anyway, i will investigate how this might be improved.

@AugustusKling , how are you calling the test?
like `yarn run test:standard` ?

**Comment by AugustusKling** [comment]
The description of this PR is confusing as it turned out the command `yarn run test:lint` does call the TypeScript compiler instead of ESLint.

When I do `yarn run cs-fix` to actually call ESLint, it still errors due to the wrong TypeScript version. Output of ESLint which is followed by bogus errors:
```
=============

WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=4.3.5 <5.4.0

YOUR TYPESCRIPT VERSION: 5.6.2

Please only submit bug reports when using the officially supported version.

=============
```

I see this behavior when running in the latest Docker container with Node 22 as in `docker run -v .:/src -it node:22 bash`. I have not tested other environments.

Regardless, switching the TypeScript version fixes the issue for me.

**Comment by jkowalleck** [comment]
Thank you so much for the research and explanations.

I've tries the latest yarn version, and it makes all the errors disappear.
it probably has ships and applies the appropriate patches for typescript and has updated hooks.

----

i was curious why all these errors appeared locally, but not on CI. see https://github.com/CycloneDX/cyclonedx-node-yarn/actions/runs/11694393247/job/32567894391
turned out: CI used Yarn 4.5.1 instead of the version set in the package.json
I'll have this adjusted/fixed in CI first, so we have a proper setup for tests and discovering solutions.

**Comment by AugustusKling** [comment]
Thanks, patches by new Yarn versions for the newer TypeScript versions can explain the differences we see.

**Comment by jkowalleck** [comment]
Updating the ESlint and plugins is a bit tricky, but definitely in my radar.
I gave it low priority, since it basically works, and only affects code style - not functionality.
So i would not spend energy for this, today.

at least it all works with the latest version of yarn and its combat-patches. 
I guess you have modified your setup already, so that you were able to test it.
Feel free to push the needed changes, to fix the issue :D

**Comment by AugustusKling** [comment]
To me the required modification is to limit TypeScript to versions below 5.4.0 as suggested in this PR. This makes it compatible with ESLint and the the used Yarn version. This would fix the `test:lint` and `cs-fix` for me locally and also probably also address the bogus errors about the missing modules in PR validations in GitHub Actions / test lint. I'd not try to get the current ESLint version running with TypeScript versions for which ESLint has no official support.

As a follow-up step, we can address updating Yarn, thereafter Typescript + ESLint.

**Comment by AugustusKling** [comment]
I set @types/node to the lasted 18.x version as this Node version is listed as supported in the Node version matrix in `.github/workflows/nodejs.yml`.

My opinion is still that the `package.json` should have ranges and the `yarn.lock` needs to be committed to guarantee a stable build with pinned versions.

**Comment by jkowalleck** [comment]
> My opinion is still that the package.json should have ranges and the yarn.lock needs to be committed to guarantee a stable build with pinned versions.

Understood.
This is a topic for another pullrequest. see #198

File: .github/workflows/nodejs.yml
Comment (jkowalleck) [inline .github/workflows/nodejs.yml:1]: please revert these changes. 
they are not in the current scope.

File: package.json [hunk 1]
```diff
@@ -69,7 +69,7 @@
   "devDependencies": {
     "@types/hosted-git-info": "^3.0.5",
     "@types/mocha": "^10.0.6",
-    "@types/node": "ts5.6",
+    "@types/node": "18.11.9",
```
Comment (jkowalleck) [inline package.json:72]:
```suggestion
    "@types/node": "ts5.3",
```

File: .github/workflows/nodejs.yml
Comment (jkowalleck) [reply .github/workflows/nodejs.yml:1]: done

File: package.json (hunk 1 above)
Comment (jkowalleck) [reply package.json]: done
